### ISSUE TITLE
Experiment:  Deferred message acknowledgement

### ISSUE THREAD

**Issue by jd0-sag** [Terracotta-OSS/terracotta-apis#86]
While the experiment in #82 did a good job of describing how an active entity could re-interpret client-originating messages for its passive counterpart(s), it did add complexity to the passive entities, in the form of the write-ahead-log, and added substantial complexity to the server, in the form of re-sending internally-generated messages (required for the multiple passive case).

Ultimately, both of these problems were to address the issue that the client would not re-send the original logical operation, because it had already returned.  We can't block execution, within the server's request processing threads, so these approaches were used in order to "fake up" a kind of re-send behavior.

This experiment describes a different approach:  **deferred message acknowledgement**.

The general idea is that the existing client-server re-send support is invoked when a message from the client doesn't receive its COMPLETE acknowledgement (and return value) from the server before a reconnect event occurs (either restart or fail-over).  We already wait for the passive to ack to the active before the active will ack to the client so it should be possible to expand this delay, further.

Ideally, this means requesting that the active server doesn't ack the logical operation to the client until all passives have acked the final sub-operation back to the active.  This still doesn't require blocking in the critical path, just changing the trigger for the original ack.

There are a few ways of exposing this:

1. Create a new service which allows the active entity to directly handle the ack for the message (asks it to become manual - which gives the entity an ack token it can activate, in the future - then activates the ack token once the logical operation is complete)
1. Extend the existing EntityMessenger service interface so that it is possible to tell it that the message it is sending should actually *inherit* the ack state of the original message

(1) is very powerful (it would be possible to implement server-side synchronization primitives without any ClientCommunicator messaging or worrying about cross-network races) but has some concerns which make it a non-starter:

- How would the ack token activation decision be made?  The sub-operation still completes without any way for the active to know this.  Additionally, it would need to be running at that time, but we want it to already have returned
- This opens the door for creating cluster-wide deadlocks and other problems as a result of entity-level software bugs.  If an ack token is dropped, that client will never recover.

(2) is more tightly constrained and describes something more literally in line with our specific intent, in this situation.  Explicitly stating that the ack to the client should happen in response to the ack for a specific active-generated message, instead of the currently executing message, means that all of the ack logic can still be handled by the platform, with no additional tracking in the entity.  The fact that the description of how the acks must be handled is still kept largely abstract means that the platform has a high degree of freedom in terms of how it interprets this requirement, allowing for freedom in future improvements.

In terms of the other requirements mentioned by #82:

- Write-ahead-log:  This is no longer required.  It existed purely to "fake up" the data which couldn't be re-sent, on fail-over.  Now that the ack is being extended, the client will re-send the original message, directly.  This is superior to tracking the data in the write-ahead-log (same content but we no know who sent the message)
- Re-send active-generated message for multiple passive case:  The re-send from the client will address this concern since all the passives will receive the original message, again.  Active-generated messages go back to their previous delivery rules where there is no re-send

**Comment by jd0-sag** [comment]
There are 2 different ways of handling this change to the acknowledgement system:

1. Server-side:  We don't send the COMPLETE ack or return value until the message which has become responsible for the ack has finished running on every server in the stripe
1. Client-side:  We send the COMPLETE ack, where we currently do, but don't return control to the client until a RETIRED ack comes in, when the last message associated with the client-originating message has run on every server in the stripe

Note:  In either case, the value returned to the client code is the return value of the initial message, not the one which caused the ack.

(1) will probably be the first approach taken in the experiment, as fewer components need to know about the change.  In this case, the return value is held, on the server, until the final message has been acked by the entire stripe and then it is returned.  If there is a fail-over, the client will re-send the message to the new active, not knowing that it ever even partially completed.

(2) is a slightly more complex change but does open the door to some other optimizations which could make it more appealing.  In this case, the return value is sent to the client, as it currently is, but the client still waits for the RETIRE ack (new), instead of immediately returning control.  In the case of a fail-over, the client acks will be reset and sees a new COMPLETE ack, and return value, while waiting for the RETIRE.

While (1) is quite simple, it could incur additional messaging latency in the cases where we only need to know that the message *will* run, everywhere, but its return value can't change.

Considered in the context of (2), it would be possible to expose the RETIRE ack to the client.  If a message could have a meaningfully different result, on fail-over, the client-side would want to wait for RETIRE.  If the message had no return value or any change in the value was considered no more correct than the initial result, the client would only wait for COMPLETE.

In this case of only waiting for COMPLETE, the key thing to realize is that the client will still re-send not-RETIRED messages, addressing the extra complexity of #82 where much of the extra entity-side logic and special active-created message re-send logic would be required specifically to get this information which the client can easily re-send.

**Comment by jd0-sag** [comment]
Another concern which may come about from this approach is regarding the order of write-back to the client, on COMPLETE.

Currently, the COMPLETE for a sequence of invokes on the same concurrency key are all issued in the order they were executed.  Now that this final ack might be completed by a different invoke, that pushes it out of the strict order.  This could cause problems on re-send, where a sequence of re-sent invokes may now contain gaps.  This would violate our current statement that, although re-sends may cause the same invoke to be run more than once, it will always be run in the same relative order.  These gaps in the order would constitute a violation of this statement.

Forcing the messages awaiting COMPLETE into a sort of "re-order buffer" would address this problem but slow down the system as no client could receive the COMPLETE message until any of these deferred acknowledgement messages before it had completed.

Proposal (2), from above, may alleviate this latency in the relevant parts of this problem.  In this case, the COMPLETE messages could be returned, as soon as possible, but only the RETIRE acks would be put into the re-order buffer.  Most client-side messages which either don't return a value or wouldn't return a different value, on re-send, could only wait on the COMPLETE message.  This way, the common-case of the client-side calls wouldn't see this additional latency but it would allow the complete picture to be re-sent, on fail-over or restart.

**Comment by jd0-sag** [comment]
More work on the prototype, today.  I created the RETIRE ack and it is now being sent through from the server (in all cases, for now).

Next step is to manage the tracking of when to fire these acks and add the EntityMessenger support for delegating this responsibility.

Some other questions that this brought up, regarding ack handling:

1. Do we want to avoid sending the RETIRE acks, on the passive?  They don't have any real meaning there.
1. How do we want to expose this, on the client?  The obvious answer is the same way we handle other acks (block on invoke() until the set of requested acks has been seen) but it probably makes more sense to block the get() in cases where we care about RETIRE.

**Comment by jd0-sag** [comment]
While this prototype will most likely depend on, and send, the RETIRE acks in all cases, I think that the only logically required case is INVOKE_ACTION messages from client-active.

This is because the other message types are never exposed to the active entity and there is no way to exploit the difference between COMPLETE/RETIRE, on the passive (it also wouldn't mean the same thing since active->passive does not re-send).

The reason to avoid specializing this, for now, is that the code around messages is quite generalized.  There may be a reason to specialize this, further, to not only clean up this issue, but also because the other message types are often handled in different ways, defined by where in the code they are processed.

**Comment by jd0-sag** [comment]
This experiment is now complete.  Branches (note that these are only prototype quality):
https://github.com/jd0-sag/terracotta-apis/tree/experiment_RetireAck
https://github.com/jd0-sag/terracotta-core/tree/experiment_RetireAck
https://github.com/jd0-sag/terracotta-platform/tree/experiment_RetireAck

Specific concerns to keep in mind if we are to proceed with this approach:

1. How to identify the hand-off between "defer retirement to the completion of this message I am about to handle" and "I am now about to run this message".  As it stands, the message is serialized and deserialized between these operations and we have no out-of-band data which could identify it.  This requires either documenting that EntityMessage types must implement a reasonable hashCode() and equals() method or changing the way that message queuing is done to allow us to pass the instance or some out-of-band data into the ProcessTransactionHandler.  In either case, we need some way of determining that it is the same message appearing in the 2 places to build the relationship.
1. Extra care now has to be put into making sure that the retirements happen, in-order, within a given concurrency key.  In the past, when we only waited for complete, this was an implicit side-effect of the operation:  the messages were run in-order so they always completed in-order.  Now, a message which has deferred its retirement must additionally block the retirement of all following messages in its concurrency key.  Otherwise, the re-sent message stream will have gaps, meaning that the re-sent messages won't maintain order.
1. While it is possible to "skip stone" the invokes (that is, continue to defer each message onto the next), this will block retirement on the entire concurrency key so we may need to directly act to prohibit this (only allowing client-originated messages to defer, for example).

Some thoughts on optimizing the common-cases:

1. The RETIRE acknowledgement is only used for client-server invoke messages but we now have to pass it, everywhere.  It may be worth specializing something about its use or the message types to avoid this redundant message in the vast majority of cases.
1. Even in the invoke path, the case of deferred retirement is still not the common case so we may need to optimize the case which avoids the extra overhead of how the deferred retirements are tracked
